### PR TITLE
Add use cases

### PR DESCRIPTION
**Pull request by tomoyukilabs**
This PR adds use cases to the WebNN draft. The use cases I've added are copied from 
https://github.com/webmachinelearning/meetings/issues/1#issuecomment-430846472 as a starting point. We will capture [suggestions and contributions from F2F discussion](https://www.w3.org/2018/10/26-webmachinelearning-minutes.html#x02) during review of this PR before merging it.

### CODE REVIEW

**Comment by anssiko** [comment]
@tomoyukilabs, thank you!

All - please review the use cases and note this PR is just a starting point and will change based on your review comments. Please suggest new use cases, clarify existing ones, rewording, de-scoping some etc.

As a guideline for reviewers, use cases are generally more impactful if implementation feasibility can be demonstrated via a proof-of-concept, mapping to platform APIs as a "reality check", or similar. 

Our [charter](https://webmachinelearning.github.io/charter/) sets the following expectation we should reflect the use cases against:

>The APIs in scope of this group will not be tied to any particular platform and will be implementable on top of existing major platform APIs, such as Android Neural Networks API, Windows DirectML, and macOS/iOS Metal Performance Shaders and Basic Neural Network Subroutines.

Reviews from folks with close familiarity with one or more of the said platform APIs very welcome.

We also have @huningxin's [API native mapping table](https://github.com/intel/webml-polyfill/blob/master/docs/native_mapping.md) at our disposal.

File: index.bs [hunk 1]
```diff
@@ -18,5 +18,160 @@ Introduction here.
 Use cases {#usecases}
 =====================
 
-Use cases here.
+## High-Level Use Cases ## {#usecases-highlevel}
 
+This section illustrates application-level use cases for the Web Machine
+Learning API (WebML API). All applications in those use cases can be built on
```
[the remark this answers is not in the document]
Comment (tomoyukilabs) [reply index.bs]: The former example looks good to me. Thanks!

File: index.bs [hunk 2]
```diff
@@ -18,5 +18,160 @@ Introduction here.
...
+top of pre-trained deep neural network (DNN) models.
+
+### Person Detection ### {#usecase-person-detection}
+
+A user is browsing a social media site and wishes to take a photo and upload it
+to the site. Before the photo is uploaded, the site runs [[SSD]] or [[YOLO]] on
```
[the remark this answers is not in the document]
Comment (tomoyukilabs) [reply index.bs]: LGTM

File: index.bs [hunk 3]
```diff
@@ -18,5 +18,160 @@ Introduction here.
...
+the WebML API to detect regions that include persons so that the user can filter
+and de-personalize irrelevant persons on it.
+
+### Skeleton Detecton ### {#usecase-skeleton-detection}
+
+A user opens a web application that continuously captures her body with her
+smartphone's camera. The web application extracts her skeleton by running
+[[PoseNet]] on the WebML API to recognize her gesture or body language. When she
```
[the remark this answers is not in the document]
Comment (tomoyukilabs) [reply index.bs]: LGTM

File: index.bs [hunk 4]
```diff
@@ -18,5 +18,160 @@ Introduction here.
...
+strikes a specified pose like raising a hand, the web application automatically
+takes a photo and upload it to an online storage.
+
+### Random Image Generation ### {#usecase-image-generation}
+
+A user wishes to make her new account and looks for a new icon image. When she
+clicks a "Generate" button on the webpage for creating an account, the webpage
+runs a generator model of generative adversarial network (GAN) for icon
+synthesis [[LogoSynthesis]] on the WebML API. She can repeat random icon
```
[the remark this answers is not in the document]
Comment (tomoyukilabs) [reply index.bs]: I agree.

File: index.bs [hunk 5]
```diff
@@ -18,5 +18,160 @@ Introduction here.
...
+generation until she finds her favorite one.
+
+## Low-Level Use Cases ## {#usecases-lowlevel}
+
+This section collects API-level use cases for the WebML API. It is supposed that
```
[the remark this answers is not in the document]
Comment (tomoyukilabs) [reply index.bs]: I agree with your opinion, and the proposed explanation looks good to me. Thanks!

**Comment by tomoyukilabs** [comment]
@anssiko Many thanks for your reviewing. I have followed your comments and updated this draft. PTAL.

**Comment by tomoyukilabs** [comment]
Currently, all application-level use cases in this PR are about CNN-based image processing. I guess that we can find any other use cases, e.g. audio, text, sensor data, etc.

**Comment by anssiko** [comment]
@tomoyukilabs, thanks for incorporating the suggestions, LGTM.

Before we consider merging, I'd like to get additional 2-3 reviews from the group, and optimally contributions for 1-2 application-level use cases that do not involve image processing.

**Comment by gregwhitworth** [comment]
@tomoyukilabs Thank you so much for taking the time to submit this PR. I agree that a text based use case would be valuable.

**Comment by huningxin** [comment]
@tomoyukilabs , thanks much for putting together this PR!

During the TPAC F2F meeting, folks were also interested in the background removal/replacement for video conference. So it would be good to add the **scene segmentation** use case, for example [[Deeplab V3+](https://github.com/tensorflow/models/tree/master/research/deeplab)] or [[Mask R-CNN](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/instance_segmentation.md)].

Other vision based use cases could include **super resolution** e.g.  [[SRGAN](https://github.com/tensorlayer/srgan)], **style transfer** e.g. [[Fast Style Transfer](https://github.com/lengstrom/fast-style-transfer)], **face analysis** e.g. [[DeepFace](https://github.com/RiweiChen/DeepFace)] and **face recognition** e.g. [[FaceNet](https://github.com/davidsandberg/facenet)]. Basically the are based on Convolutional Neural Networks (CNN).

Some text based use cases could be **machine translation** e.g. [[GNMT](https://github.com/tensorflow/nmt)] or [[OpenNMT](https://github.com/OpenNMT)], **sentiment analysis**  e.g. [[DeepMoji](https://github.com/bfelbo/DeepMoji)], **speech recognition** e.g. [[Deep Speech](https://github.com/mozilla/DeepSpeech)], **text to speech** e.g. [[Deep Voice](https://github.com/r9y9/deepvoice3_pytorch)], **image captioning** e.g. [[im2txt](https://github.com/tensorflow/models/tree/master/research/im2txt)] and **video summarization** e.g. [[Video-Summarization-with-LSTM](https://github.com/kezhang-cs/Video-Summarization-with-LSTM)]. They are usually based on Recurrent Neural Networks (RNN).

File: index.bs [hunk 6]
```diff
@@ -18,5 +18,166 @@ Introduction here.
...
+### Custom Layer ### {#usecase-custom-layer}
+
+A web application developer wants to run a DNN model on the WebNN API. However,
+she has found that some of activation functions like [[LeakyReLU]], [[ELU]],
+etc. are not included in the WebNN API. So she constructs custom layers of the
+additional activation functions on top of the WebNN API. Note that the scope of
+custom layers may include convolution, normalization, etc. as well as
+activation.
+
+### Network Concatenation ### {#usecase-network-concat}
```
Comment (huningxin) [inline index.bs]: As mentioned in the TPAC F2F meeting, this looks like a training use case. As training is out of scope of current charter, would it be better to add this in the future?

Comment (tomoyukilabs) [reply index.bs]: Correct, but not limited to training. Possible detailed examples are:

- The web app downloads convolutional layer weights of MobileNetV1/V2 from CDN and weights of fully-connected layers made by transfer learning from her own web site
- The web app downloads complete weights of MobileNetV1/V2, and then partially update fully-connected layers by downloading fine-tuned weights

Anyway, the current description seems to suggest the use case of training, as you pointed out. I'll update those sentences so that they clearly indicate a use case of client-side partial update based on fine tuning or transfer learning.

**Comment by tomoyukilabs** [comment]
@huningxin All use cases you have suggested including text-based ones look great to me. Many thanks!

@anssiko Is it okay to add those use cases to this PR?

**Comment by anssiko** [comment]
@tomoyukilabs, yes please add. Similarly to the initial list of use cases, the group is expected to review any proposed additions and doing that is easier using the PR review facilities.

**Comment by anssiko** [comment]
@huningxin, thanks for the great contribution!

Since accessibility is a key to the [W3C's mission](https://www.w3.org/Consortium/mission), maybe it's worth noting the accessibility benefit in connection with the image captioning use case [[im2txt](https://github.com/tensorflow/models/tree/master/research/im2txt)]. Being able to add image descriptions automatically greatly improves web accessibility. As we know, only a small fraction of images on the Web have been properly annotated.

**Comment by tomoyukilabs** [comment]
Due to my business trip until the end of November, I'll start updating this PR as soon as possible after coming back to Tokyo. Thanks for your patience.

**Comment by tomoyukilabs** [comment]
I've updated this PR. PTAL.

- The use cases proposed by @huningxin in https://github.com/webmachinelearning/webnn/pull/2#issuecomment-440193296 are added
- Regarding the use case of GAN, image generation is replaced with super resolution, i.e. SRGAN
- Person detection and skeleton detection are revised as the use cases of video conferencing
- According to https://github.com/webmachinelearning/webnn/pull/2#pullrequestreview-177037342, model concatenation is modified as the use case of fine tuning.

File: index.bs [hunk 7]
```diff
@@ -18,5 +18,287 @@ Introduction here.
...
+Therefore, the application downloads both partial model files at first and
+concatenates them into a single model. When the model is updated, the
+application downloads fine-tuned part of the model and replace only the
+fully-connected layers with it.
+
+### Performance Adaptation ### {#usecase-perf-adapt}
+
+A web application developer has a concern about performance of her DNN model on
+mobile devices. She has confirmed that the model runs too slow on mobile devices
+which does not have GPU acceleration. So her web application refers to the WebNN
```
Comment (anssiko) [inline index.bs]: s/does/do/
s/So/To address this issue,/

File: index.bs [hunk 8]
```diff
@@ -18,5 +18,287 @@ Introduction here.
...
+API) and the low-level details exposed through the WebNN API are abstracted out
+from typical web developers. However, it is also expected that web developers
+with specific interest and competence in Machine Learning will want to interface
+with the WebNN API directly instead of a higher-level ML framework.
+
+### Custom Layer ### {#usecase-custom-layer}
+
+A web application developer wants to run a DNN model on the WebNN API. However,
+she has found that some of activation functions like [[LeakyReLU]], [[ELU]],
+etc. are not included in the WebNN API. So she constructs custom layers of the
```
Comment (anssiko) [inline index.bs]: s/So/To address this issue,/

File: index.bs [hunk 9]
```diff
@@ -18,5 +18,287 @@ Introduction here.
...
+using a web-based video conferencing application. The application detects faces
+of participants by using object detection (for example, using object detection
+approaches such as [[SSD]]) and checks whether each face was present at the
+previous meeting or not by running a machine learning model such as [[FaceNet]],
+which verifies whether two faces would be identical or not.
+
+### Super Resolution ### {#usecase-super-resolution}
+
+A web-based video conferencing is receiving a video stream from its peer, but
+the resolution of the video becomes lower due to network congestion. So the
```
Comment (anssiko) [inline index.bs]: /So/To prevent degradation of the perceived video quality,/

File: index.bs [hunk 10]
```diff
@@ -18,5 +18,287 @@ Introduction here.
...
+input frame that include persons.
+
+When she comes back, the application automatically detects her and notifies
+other online users that she is active now.
+
+### Semantic Segmentation ### {#usecase-segmentation}
+
+A user joins a teleconference via a web-based video conferencing application
+from her room. However, she does not wish that her room is visible on the
+screen. So the application runs a machine learning model such as [[DeepLabv3+]]
```
Comment (anssiko) [inline index.bs]: /So/To protect the privacy of the other people and the surroundings,/

File: index.bs [hunk 11]
```diff
@@ -18,5 +18,287 @@ Introduction here.
...
+such as [[SSD]] or [[YOLO]] that use a single DNN) to detect regions in a camera
+input frame that include persons.
+
+When she comes back, the application automatically detects her and notifies
+other online users that she is active now.
+
+### Semantic Segmentation ### {#usecase-segmentation}
+
+A user joins a teleconference via a web-based video conferencing application
+from her room. However, she does not wish that her room is visible on the
```
Comment (anssiko) [inline index.bs]: s/room is/room and people in the background are/

File: index.bs [hunk 12]
```diff
@@ -18,5 +18,287 @@ Introduction here.
...
+application downloads fine-tuned part of the model and replace only the
+fully-connected layers with it.
+
+### Performance Adaptation ### {#usecase-perf-adapt}
+
+A web application developer has a concern about performance of her DNN model on
+mobile devices. She has confirmed that the model runs too slow on mobile devices
+which does not have GPU acceleration. So her web application refers to the WebNN
+API to confirm whether acceleration is available or not, so that the application
+can display the warning for devices without acceleration.
```
Comment (anssiko) [inline index.bs]: s/the/a/

File: index.bs [hunk 13]
```diff
@@ -18,5 +18,287 @@ Introduction here.
...
+
+After several weeks, she has developed a tiny DNN model that can even run on
+CPU. So she modifies the application so that the application loads the tiny
```
Comment (anssiko) [inline index.bs]: s/So/In order to accommodate for that,"

**Comment by tomoyukilabs** [comment]
@anssiko Thanks for your review. I've revised this PR. PTAL.

**Comment by anssiko** [comment]
@tomoyukilabs, thanks, LGTM.

All - We'll review this PR during our [13 December 2018 teleconference](https://github.com/webmachinelearning/meetings/blob/master/telcons/2018-12-13-agenda.md) to assess consensus whether this set of use cases represents a good starting point for the initial API design.

See also the [HTML preview of these use cases](https://pr-preview.s3.amazonaws.com/tomoyukilabs/webnn/pull/2.html#usecases). All feedback welcome.

**Comment by zhiqiangyu** [comment]
These use cases look good.  I like to propose one more case as below, mainly for web shopping scenario,  could you please take a look?  Any comment is welcome.  Thanks.

Facial Features Detection:
A web-base shopping application detects user's facial features (e.g. the detailed information of eyes/nose/mouth/lips/etc),  and enable user to perform beautify try-on simulations,such as wear the glasses,  perform lipstick make-up, etc.  An example could be found here: http://modiface.com/.   Further more,  this kind of capability can be also extended to more scenarios like human face modelling, emotion analysis, etc.

**Comment by anssiko** [comment]
@zhiqiangyu, thanks! Facial features detection is indeed an important step in various facial analysis tasks, out of which we currently list face recognition and emotion analysis (could also be used for drowsiness detection in the person detection use case! 😴)

@tomoyukilabs @huningxin, how would you suggest we integrate facial features detection given it is a key step in many facial analysis tasks? Also, would it help to mention some commonly used facial landmark detection approaches?

If we'd generalize, could say facial features (or landmark) detection enables a number of use cases in HCI, entertainment (incl. shopping), medical, security surveillance, and more.

**Comment by tomoyukilabs** [comment]
@zhiqiangyu Thanks. That use case looks good to me!

@anssiko @huningxin  I have added a couple of use cases related to facial characteristics. PTAL.

- *Facial Landmark Detection*: [Face Alignment Network](https://github.com/1adrianb/2D-and-3D-face-alignment) can detect 2D and 3D facial landmarks in detail.
- *Style Transfer*: This is one of the use cases that @huningxin has proposed in 
https://github.com/webmachinelearning/webnn/pull/2#issuecomment-440193296. [The Contextual Loss](https://github.com/roimehrez/contextualLoss) and [PairedCycleGAN](https://research.adobe.com/project/makeupgan-makeup-transfer-via-cycle-consistent-adversarial-networks/) can transfer makeup style from one facial image to another.
